### PR TITLE
Apiserver requestheader security fix

### LINKED ISSUE #119267: NCC-E003660-F9W: Common Certificate Authority Possible for Client CA and Request
### NCC-E003660-F9W: Common Certificate Authority Possible for Client CA and Request
Header CA
This issue was reported in the [Kubernetes 1.24 Security Audit Report](https://github.com/kubernetes/sig-security/blob/main/sig-security-external-audit/security-audit-2021-2022/findings/Kubernetes%20v1.24%20Final%20Report.pdf)

**Impact**
A misconfiguration of the API server could lead to a situation where privilege escalation for
any authenticated user to cluster admin is possible.

**Description**
The Kubernetes API server allows a certificate authority (CA) to be specified (using the `--
client-ca-file flag`) which will be used to verify client certificates under the familiar X.
509 client certificate authentication scheme6. Another CA can be specified (using `--
requestheader-client-ca-file`) that will be used to verify requests authenticated using
the request header scheme (described in the documentation as Authenticating Proxy). If
these two CAs are the same, and the `--requestheader-allowed-names` flag is not used to
specify particular certificate common names that are permitted for request header
authentication, it is possible for any holder of a client certificate from the shared CA to
trivially authenticate to the API server as any user, by using their certificate to authenticate
a request using the request header scheme. While it is not intended that these CAs be the
same, and common Kubernetes deployment methods do not configure the deployment in
this way, it is possible that administrators could be unaware of this pitfall.

The Kubernetes documentation does include some advice against sharing CAs, but is not
explicit about the potentially serious security impact of this specific case.

For an example, please see the original finding on page 18 of the [Kubernetes 1.24 Security Audit Report](https://github.com/kubernetes/sig-security/blob/main/sig-security-external-audit/security-audit-2021-2022/findings/Kubernetes%20v1.24%20Final%20Report.pdf)

**Recommendation**
The Kubernetes API server should reject any configuration where `--client-ca-file` and
`--requestheader-client-ca-file` are set to the same value, unless `--requestheaderallowed-
names` has been set to specific value(s).

It may also be advisable to update the Kubernetes documentation to be more explicit about
the specific risks of sharing certificate authorities. For example, using the same CA for X.
509 client certificates for etcd authentication would allow any authenticated user to
directly access etcd (although this scenario is considered less likely than sharing of the
client CA and request header CA).

**Component**
kube-apiserver

**Anything else we need to know?**
See umbrella issue [#118980](https://github.com/kubernetes/kubernetes/issues/118980) for current status of all issues created from these findings.
The vendor gave this issue an ID of NCC-E003660-F9W and it was finding 5 of the report under the kube-apiserver section.
The vendor considers this issue Medium Risk, High Impact, Low Exploitability
To view the original finding, begin on page 18 of the [Kubernetes 1.24 Security Audit Report](https://github.com/kubernetes/sig-security/blob/main/sig-security-external-audit/security-audit-2021-2022/findings/Kubernetes%20v1.24%20Final%20Report.pdf)

**Test Environment**
Kubernetes 1.24.3

### PR DESCRIPTION
**Pull request by ballista01**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
This PR introduces a validation check at server startup for the Kubernetes API server configuration. It ensures that when client-ca-file and requestheader-client-ca-file are configured with overlapping certificate authorities, requestheader-allowed-names must also be specified. This change aims to prevent a potential security risk where an attacker could exploit the request header authentication method when overlapping CAs are used without proper restrictions on allowed names.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/119267

#### Special notes for your reviewer:
This validation enhances the security by avoiding a misconfiguration that could lead to significant vulnerabilities. It is a preventive measure against unauthorized privilege escalation.

#### Does this PR introduce a user-facing change?

```release-note
Changed the API server code to abort when detecting an insecure combination of trust anchors (if the same trust anchor is used to verify general client connections and to verify connections that are trusted to set a client identity). The API server also logs the cause of the abort.

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sftim** [comment]
Changelog suggestion

```diff
-Since `requestheader-allowed-names` is now required in Kubernetes API server configuration when `client-ca-file` and `requestheader-client-ca-file` have overlapping CAs. The insecure configuration will throw an error, which could be found in the apiserver log.
+Changed the API server code to abort when detecting an insecure combination of trust anchors
+(if the same trust anchor is used to verify general client connections and to verify connections
+that are trusted to set a client identity). The API server also logs the cause of the abort.
```

**Comment by ballista01** [comment]
@sftim Thanks for the suggestion! I'll update the changelog.

File: pkg/kubeapiserver/options/authentication.go [hunk 1]
```diff
@@ -269,6 +269,25 @@ func (o *BuiltInAuthenticationOptions) Validate() []error {
...
 		allErrors = append(allErrors, o.RequestHeader.Validate()...)
+		if o.ClientCert != nil {
+			// Enforce 'requestheader-allowed-names' check when 'requestheader-client-ca-file'
+			// and 'client-ca-file' are the same and not empty, to prevent unauthorized privilege escalation.
+			clientCA := strings.TrimSpace(o.ClientCert.ClientCA)
+			requestHeaderClientCA := strings.TrimSpace(o.RequestHeader.ClientCAFile)
+			if len(clientCA) != 0 && clientCA == requestHeaderClientCA {
+				numValidNames := 0
+				for _, allowedName := range o.RequestHeader.AllowedNames {
+					if len(strings.TrimSpace(allowedName)) != 0 {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go:280]: no need to TrimSpace, that's already guarded by checkForWhiteSpaceOnly, can just check `len(o.RequestHeader.AllowedNames) == 0`

File: pkg/kubeapiserver/options/authentication.go [hunk 2]
```diff
@@ -269,6 +269,25 @@ func (o *BuiltInAuthenticationOptions) Validate() []error {
 
 	if o.RequestHeader != nil {
 		allErrors = append(allErrors, o.RequestHeader.Validate()...)
+		if o.ClientCert != nil {
+			// Enforce 'requestheader-allowed-names' check when 'requestheader-client-ca-file'
+			// and 'client-ca-file' are the same and not empty, to prevent unauthorized privilege escalation.
+			clientCA := strings.TrimSpace(o.ClientCert.ClientCA)
+			requestHeaderClientCA := strings.TrimSpace(o.RequestHeader.ClientCAFile)
+			if len(clientCA) != 0 && clientCA == requestHeaderClientCA {
```
Comment (liggitt) [inline pkg/kubeapiserver/options/authentication.go:277]: comparing the filenames is not really what we want, this needs to load both CA pools and check if any CA is in both client CA and requestheader CA list:

```go
if len(o.RequestHeader.AllowedNames) == 0 && o.ClientCert != nil && len(o.ClientCert.ClientCA) > 0 && len(o.RequestHeader.ClientCAFile) > 0 {
  clientCACerts, _ := cert.CertsFromFile(o.ClientCert.ClientCA)
  requestHeaderCACerts, _ := cert.CertsFromFile(o.RequestHeader.ClientCAFile)
  ... check if any cert is in both clientCACerts and requestHeaderCACerts
}
```

**Pull request closed without merging**